File: src/DiagToolsValidationRunner/DiagToolsValidationRunner.Core/Functionality/Tracing.cs
using System.Collections;

namespace DiagToolsValidationRunner.Core.Functionality
{
    public class PerfCollect
    {
        private readonly string PerfCollectURL = 
            "https://raw.githubusercontent.com/microsoft/perfview/main/src/perfcollect/perfcollect";
        private readonly string PerfCollectPath;

        public PerfCollect(string? perfcollectPath = null,
                           bool installPrerequisites = false)
        {
            if (!OperatingSystem.IsLinux())
            {
                throw new Exception($"{nameof(PerfCollect)}: perfcollect only works on Linux");
            }

            Dictionary<string, string> env = new();
            foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
            {
                env[de!.Key!.ToString()!] = de!.Value!.ToString()!;
            }
            // Download perfcollect
            if (String.IsNullOrEmpty(perfcollectPath))
            {
                PerfCollectPath = Path.GetTempFileName();
            }
            else
            {
                PerfCollectPath = perfcollectPath;
            }

            Task.WaitAll(Utilities.Download(PerfCollectURL, PerfCollectPath));

            // Make the script executable
            CommandInvoker executableInvoker = new("chmod",
                                                   $"+x {PerfCollectPath}",
                                                   env,
                                                   "",
                                                   false);
            executableInvoker.WaitForResult();

            if (installPrerequisites)
            {
                CommandInvoker installInvoker = new("/bin/bash",
                                                    $"{PerfCollectPath} install",
                                                    env,
                                                    "",
                                                    false);
                installInvoker.WaitForResult();
            }
        }

        public CommandInvoker CollectTraceForSecs(Dictionary<string, string> env,
                                                  string tracePath,
                                                  int collectSecs,
                                                  string workingDirectory = "",
                                                  bool redirectStdOutErr = false,
                                                  bool silent = false)
        {
            return new("/bin/bash",
                       $"{PerfCollectPath} collect {tracePath} -collectsec {collectSecs}",
                       env,
                       workingDirectory,
                       redirectStdOutErr,
                       silent);
        }
    }
}

File: src/DiagToolsTestRunner/core_functionality/tracing.py
'''Tracing tool for .NET app
'''

import os

from core_functionality import common
from core_functionality.cli import CommandInvoker

class PerfCollect:
    '''Tracing with perfcollect.
    '''
    def __init__(self, perfcollect_path: str, install_prerequisites: bool=False):
        '''
        :param perfcollect_path: perfcollect path
        :param install_prerequisites: whether to install prerequisites
        '''
        if not common.rid_os_name().startswith('linux'):
            raise OSError('perfcollect is only supported on Linux')

        self.__perfcollect_download_link = \
            'https://raw.githubusercontent.com/microsoft/perfview/main/src/perfcollect/perfcollect'
        self.__perfcollect_path = perfcollect_path

        common.http_download(self.__perfcollect_path, self.__perfcollect_download_link)

        enable_execute_args = ['chmod', '+x', self.__perfcollect_path]
        with CommandInvoker(enable_execute_args, os.environ, silent=False) as ci:
            ci.communicate()

        if install_prerequisites:
            install_args = [
                '/bin/bash',
                self.__perfcollect_path,
                'install'
            ]
            with CommandInvoker(install_args, os.environ, silent=False) as ci:
                ci.communicate()

    @property
    def perfcollect_path(self):
        '''Get perfcollect path.
        '''
        return self.__perfcollect_path

    def collect_trace_for_secs(self,
                               trace_path: str,
                               collect_secs: int,
                               redirect_std_out_err: bool=False,
                               silent: bool=False):
        '''Tracing for specified time.

        :param trace_path: path of trace file
        :param collect_secs: time span for tracing(seconds)
        :param redirect_std_out_err: whether to redirect stardard output and err
        :param silent: whether to suppress console output
        '''
        args = [
            '/bin/bash',
            self.__perfcollect_path,
            'collect',
            trace_path,
            '-collectsec',
            str(collect_secs)
        ]
        with CommandInvoker(
            args,
            os.environ,
            redirect_std_out_err=redirect_std_out_err,
            silent=silent
        ) as invoker:
            invoker.communicate()
            return invoker
